File: Tools/gen_graybox_terrain_obj.py
"""Generate a continuous graybox expedition terrain mesh (OBJ) with POI height variation and road corridors."""
from __future__ import annotations

import math
import struct
from pathlib import Path

# World coverage (cm), matching graybox POI ring
X_MIN, X_MAX = -12000.0, 12000.0
Y_MIN, Y_MAX = -2000.0, 17000.0
# ~100 cm spacing
STEP = 100.0

# POI height targets (surface Z cm)
POIS = [
    # (x, y, height, radius, falloff softness)
    (0.0, 1500.0, 0.0, 4500.0, 1.2),       # spawn/hub basin
    (-8000.0, 3000.0, -250.0, 2800.0, 1.1),  # scan valley
    (8000.0, 3000.0, 350.0, 2800.0, 1.1),    # harvest rise
    (0.0, 9000.0, 450.0, 2800.0, 1.15),      # ambush ridge
    (0.0, 14200.0, 650.0, 3200.0, 1.2),      # base plateau
]

# Road polylines: list of (x,y) waypoints; carved flatter/slightly raised path
ROADS = [
    [(0.0, 3000.0), (-8000.0, 3000.0)],          # hub -> scan
    [(0.0, 3000.0), (8000.0, 3000.0)],           # hub -> harvest
    [(0.0, 3000.0), (0.0, 9000.0)],              # hub -> ambush
    [(0.0, 9000.0), (0.0, 14200.0)],             # ambush -> base
]
ROAD_HALF_WIDTH = 700.0
ROAD_BLEND = 350.0
ROAD_RAISE = 25.0


def clamp(v: float, lo: float, hi: float) -> float:
    return lo if v < lo else hi if v > hi else v


def smoothstep(edge0: float, edge1: float, x: float) -> float:
    t = clamp((x - edge0) / (edge1 - edge0), 0.0, 1.0)
    return t * t * (3.0 - 2.0 * t)


def poi_height(x: float, y: float) -> float:
    # Weighted blend of POI targets over a mild base undulation
    base = 40.0 * math.sin(x * 0.00035) * math.cos(y * 0.00028)
    num = 0.0
    den = 0.0
    for px, py, h, radius, soft in POIS:
        dx = x - px
        dy = y - py
        d = math.sqrt(dx * dx + dy * dy)
        # Inverse-distance weight with soft falloff outside radius
        w = 1.0 / (1.0 + (d / max(radius, 1.0)) ** (2.0 * soft))
        num += w * h
        den += w
    return base + (num / den if den > 1e-6 else 0.0)


def dist_to_segment(px: float, py: float, ax: float, ay: float, bx: float, by: float) -> float:
    abx, aby = bx - ax, by - ay
    apx, apy = px - ax, py - ay
    ab2 = abx * abx + aby * aby
    if ab2 < 1e-6:
        return math.sqrt(apx * apx + apy * apy)
    t = clamp((apx * abx + apy * aby) / ab2, 0.0, 1.0)
    cx, cy = ax + t * abx, ay + t * aby
    dx, dy = px - cx, py - cy
    return math.sqrt(dx * dx + dy * dy)


def road_influence(x: float, y: float) -> tuple[float, float]:
    """Return (influence 0..1, target height along road)."""
    best_d = 1e18
    best_h = 0.0
    for poly in ROADS:
        for i in range(len(poly) - 1):
            ax, ay = poly[i]
            bx, by = poly[i + 1]
            d = dist_to_segment(x, y, ax, ay, bx, by)
            if d < best_d:
                best_d = d
                # interpolate height targets at endpoints via poi_height
                # use distance along segment for blend
                abx, aby = bx - ax, by - ay
                ab2 = abx * abx + aby * aby
                apx, apy = x - ax, y - ay
                t = 0.0 if ab2 < 1e-6 else clamp((apx * abx + apy * aby) / ab2, 0.0, 1.0)
                ha = poi_height(ax, ay)
                hb = poi_height(bx, by)
                best_h = ha + (hb - ha) * t + ROAD_RAISE
    outer = ROAD_HALF_WIDTH + ROAD_BLEND
    if best_d >= outer:
        return 0.0, best_h
    if best_d <= ROAD_HALF_WIDTH:
        return 1.0, best_h
    # blend zone
    u = 1.0 - smoothstep(ROAD_HALF_WIDTH, outer, best_d)
    return u, best_h


def sample_height(x: float, y: float) -> float:
    h = poi_height(x, y)
    infl, road_h = road_influence(x, y)
    if infl <= 0.0:
        return h
    # Flatten toward road corridor height
    return h * (1.0 - infl) + road_h * infl


def main() -> None:
    xs = int((X_MAX - X_MIN) / STEP) + 1
    ys = int((Y_MAX - Y_MIN) / STEP) + 1
    out_path = Path(__file__).resolve().parents[1] / "Content" / "Level" / "Graybox" / "SM_GrayboxExpeditionTerrain.obj"
    out_path.parent.mkdir(parents=True, exist_ok=True)

    verts: list[tuple[float, float, float]] = []
    for j in range(ys):
        y = Y_MIN + j * STEP
        for i in range(xs):
            x = X_MIN + i * STEP
            z = sample_height(x, y)
            verts.append((x, y, z))

    # Write OBJ (Unreal: X forward? Import typically Y-up or Z-up depending on options.
    # UE FBX/OBJ import defaults often treat Z-up. We'll emit Z-up: x,y horizontal, z height.
    with out_path.open("w", encoding="utf-8", newline="\n") as f:
        f.write("# GrayboxExpedition continuous terrain\n")
        f.write(f"# grid {xs} x {ys}, step {STEP}cm\n")
        for x, y, z in verts:
            f.write(f"v {x:.3f} {y:.3f} {z:.3f}\n")
        # faces (1-based), CCW when viewed from above
        for j in range(ys - 1):
            for i in range(xs - 1):
                i0 = j * xs + i + 1
                i1 = i0 + 1
                i2 = i0 + xs
                i3 = i2 + 1
                f.write(f"f {i0} {i1} {i3}\n")
                f.write(f"f {i0} {i3} {i2}\n")

    # Also write a 16-bit raw heightmap for optional Landscape import by hand
    hm_path = out_path.with_suffix(".r16")
    zmin = min(v[2] for v in verts)
    zmax = max(v[2] for v in verts)
    zrange = max(zmax - zmin, 1.0)
    with hm_path.open("wb") as f:
        for j in range(ys):
            for i in range(xs):
                z = verts[j * xs + i][2]
                u16 = int(clamp((z - zmin) / zrange, 0.0, 1.0) * 65535.0 + 0.5)
                f.write(struct.pack("<H", u16))

    meta = out_path.with_suffix(".json")
    meta.write_text(
        "{\n"
        f'  "xs": {xs},\n'
        f'  "ys": {ys},\n'
        f'  "step_cm": {STEP},\n'
        f'  "x_min": {X_MIN}, "x_max": {X_MAX},\n'
        f'  "y_min": {Y_MIN}, "y_max": {Y_MAX},\n'
        f'  "z_min": {zmin:.3f}, "z_max": {zmax:.3f},\n'
        f'  "obj": "{out_path.as_posix()}",\n'
        f'  "r16": "{hm_path.as_posix()}"\n'
        "}\n",
        encoding="utf-8",
    )
    print(f"Wrote {out_path} ({xs}x{ys} verts={len(verts)}) z=[{zmin:.1f},{zmax:.1f}]")
    print(f"Wrote {hm_path}")


if __name__ == "__main__":
    main()

File: Tools/gen_graybox_terrain_obj.js
/**
 * Generate continuous graybox expedition terrain mesh (OBJ) + r16 heightmap.
 */
const fs = require("fs");
const path = require("path");

const X_MIN = -12000,
  X_MAX = 12000;
const Y_MIN = -2000,
  Y_MAX = 17000;
const STEP = 100;

const POIS = [
  [0, 1500, 0, 4500, 1.2],
  [-8000, 3000, -250, 2800, 1.1],
  [8000, 3000, 350, 2800, 1.1],
  [0, 9000, 450, 2800, 1.15],
  [0, 14200, 650, 3200, 1.2],
];

const ROADS = [
  [
    [0, 3000],
    [-8000, 3000],
  ],
  [
    [0, 3000],
    [8000, 3000],
  ],
  [
    [0, 3000],
    [0, 9000],
  ],
  [
    [0, 9000],
    [0, 14200],
  ],
];
const ROAD_HALF_WIDTH = 700;
const ROAD_BLEND = 350;
const ROAD_RAISE = 25;

function clamp(v, lo, hi) {
  return v < lo ? lo : v > hi ? hi : v;
}
function smoothstep(e0, e1, x) {
  const t = clamp((x - e0) / (e1 - e0), 0, 1);
  return t * t * (3 - 2 * t);
}

function poiHeight(x, y) {
  const base = 40 * Math.sin(x * 0.00035) * Math.cos(y * 0.00028);
  let num = 0,
    den = 0;
  for (const [px, py, h, radius, soft] of POIS) {
    const dx = x - px,
      dy = y - py;
    const d = Math.sqrt(dx * dx + dy * dy);
    const w = 1 / (1 + Math.pow(d / Math.max(radius, 1), 2 * soft));
    num += w * h;
    den += w;
  }
  return base + (den > 1e-6 ? num / den : 0);
}

function distToSegment(px, py, ax, ay, bx, by) {
  const abx = bx - ax,
    aby = by - ay;
  const apx = px - ax,
    apy = py - ay;
  const ab2 = abx * abx + aby * aby;
  if (ab2 < 1e-6) return Math.sqrt(apx * apx + apy * apy);
  const t = clamp((apx * abx + apy * aby) / ab2, 0, 1);
  const cx = ax + t * abx,
    cy = ay + t * aby;
  const dx = px - cx,
    dy = py - cy;
  return Math.sqrt(dx * dx + dy * dy);
}

function roadInfluence(x, y) {
  let bestD = 1e18,
    bestH = 0;
  for (const poly of ROADS) {
    for (let i = 0; i < poly.length - 1; i++) {
      const [ax, ay] = poly[i];
      const [bx, by] = poly[i + 1];
      const d = distToSegment(x, y, ax, ay, bx, by);
      if (d < bestD) {
        bestD = d;
        const abx = bx - ax,
          aby = by - ay;
        const ab2 = abx * abx + aby * aby;
        const apx = x - ax,
          apy = y - ay;
        const t = ab2 < 1e-6 ? 0 : clamp((apx * abx + apy * aby) / ab2, 0, 1);
        const ha = poiHeight(ax, ay);
        const hb = poiHeight(bx, by);
        bestH = ha + (hb - ha) * t + ROAD_RAISE;
      }
    }
  }
  const outer = ROAD_HALF_WIDTH + ROAD_BLEND;
  if (bestD >= outer) return [0, bestH];
  if (bestD <= ROAD_HALF_WIDTH) return [1, bestH];
  return [1 - smoothstep(ROAD_HALF_WIDTH, outer, bestD), bestH];
}

function sampleHeight(x, y) {
  const h = poiHeight(x, y);
  const [infl, roadH] = roadInfluence(x, y);
  if (infl <= 0) return h;
  return h * (1 - infl) + roadH * infl;
}

const xs = Math.floor((X_MAX - X_MIN) / STEP) + 1;
const ys = Math.floor((Y_MAX - Y_MIN) / STEP) + 1;
const verts = new Array(xs * ys);
let zmin = Infinity,
  zmax = -Infinity;

for (let j = 0; j < ys; j++) {
  const y = Y_MIN + j * STEP;
  for (let i = 0; i < xs; i++) {
    const x = X_MIN + i * STEP;
    const z = sampleHeight(x, y);
    verts[j * xs + i] = [x, y, z];
    if (z < zmin) zmin = z;
    if (z > zmax) zmax = z;
  }
}

const outDir = path.join(__dirname, "..", "Content", "Level", "Graybox");
fs.mkdirSync(outDir, { recursive: true });
const objPath = path.join(outDir, "SM_GrayboxExpeditionTerrain.obj");

let obj = `# GrayboxExpedition continuous terrain\n# grid ${xs} x ${ys}, step ${STEP}cm\n`;
for (const [x, y, z] of verts) {
  obj += `v ${x.toFixed(3)} ${y.toFixed(3)} ${z.toFixed(3)}\n`;
}
for (let j = 0; j < ys - 1; j++) {
  for (let i = 0; i < xs - 1; i++) {
    const i0 = j * xs + i + 1;
    const i1 = i0 + 1;
    const i2 = i0 + xs;
    const i3 = i2 + 1;
    obj += `f ${i0} ${i1} ${i3}\n`;
    obj += `f ${i0} ${i3} ${i2}\n`;
  }
}
fs.writeFileSync(objPath, obj);

const r16Path = path.join(outDir, "SM_GrayboxExpeditionTerrain.r16");
const zrange = Math.max(zmax - zmin, 1);
const buf = Buffer.alloc(xs * ys * 2);
for (let n = 0; n < verts.length; n++) {
  const u16 = Math.round(clamp((verts[n][2] - zmin) / zrange, 0, 1) * 65535);
  buf.writeUInt16LE(u16, n * 2);
}
fs.writeFileSync(r16Path, buf);

fs.writeFileSync(
  path.join(outDir, "SM_GrayboxExpeditionTerrain.json"),
  JSON.stringify(
    {
      xs,
      ys,
      step_cm: STEP,
      x_min: X_MIN,
      x_max: X_MAX,
      y_min: Y_MIN,
      y_max: Y_MAX,
      z_min: zmin,
      z_max: zmax,
      obj: objPath,
      r16: r16Path,
    },
    null,
    2
  )
);

console.log(
  `Wrote ${objPath} (${xs}x${ys} verts=${verts.length}) z=[${zmin.toFixed(1)},${zmax.toFixed(1)}]`
);
